domain/src/test/java/org/oppia/domain/topic/TopicControllerTest.kt

```diff
@@ -767,13 +793,344 @@ class TopicControllerTest {
+      assertThat(topic.storyList[0].chapterList[0].chapterPlayState).isEqualTo(ChapterPlayState.NOT_STARTED)
+      assertThat(topic.storyList[0].chapterList[1].chapterPlayState).isEqualTo(ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES)
+    }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testGetTopic_recordProgress_getTopic_correctProgressFound() =
+    runBlockingTest(coroutineContext) {
+      markFractionsStory0Chapter0AsCompleted()
+      advanceUntilIdle()
+
+      topicController.getTopic(profileId1, FRACTIONS_TOPIC_ID).observeForever(mockTopicObserver)
+      advanceUntilIdle()
+
+      verifyGetTopicSucceeded()
+
+      val topic = topicResultCaptor.value.getOrThrow()
+      assertThat(topic.topicId).isEqualTo(FRACTIONS_TOPIC_ID)
+      assertThat(topic.storyList[0].chapterList[0].chapterPlayState).isEqualTo(ChapterPlayState.COMPLETED)
+      assertThat(topic.storyList[0].chapterList[1].chapterPlayState).isEqualTo(ChapterPlayState.NOT_STARTED)
+    }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testGetStory_invalidData_getStory_noResultFound() =
+    runBlockingTest(coroutineContext) {
+      topicController.getStory(profileId1, INVALID_TOPIC_ID_1, INVALID_STORY_ID_1)
+        .observeForever(mockStorySummaryObserver)
+      advanceUntilIdle()
+
+      verifyGetStoryFailed()
+    }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testGetStory_validData_withoutAnyProgress_getStorySucceedsWithCorrectProgress() =
+    runBlockingTest(coroutineContext) {
+      topicController.getStory(profileId1, FRACTIONS_TOPIC_ID, FRACTIONS_STORY_ID_0)
+        .observeForever(mockStorySummaryObserver)
+      advanceUntilIdle()
+
+      verifyGetStorySucceeded()
+
+      val storySummary = storySummaryResultCaptor.value.getOrThrow()
+      assertThat(storySummary.storyId).isEqualTo(FRACTIONS_STORY_ID_0)
+      assertThat(storySummary.chapterList[0].chapterPlayState).isEqualTo(ChapterPlayState.NOT_STARTED)
+      assertThat(storySummary.chapterList[1].chapterPlayState).isEqualTo(ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES)
+    }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testGetStory_recordProgress_getTopic_correctProgressFound() =
+    runBlockingTest(coroutineContext) {
+      markFractionsStory0Chapter0AsCompleted()
+      advanceUntilIdle()
+
+      topicController.getTopic(profileId1, FRACTIONS_TOPIC_ID).observeForever(mockTopicObserver)
+      advanceUntilIdle()
+
+      verifyRecordProgressSucceeded()
+      verifyGetTopicSucceeded()
+
+      val topic = topicResultCaptor.value.getOrThrow()
+      assertThat(topic.topicId).isEqualTo(FRACTIONS_TOPIC_ID)
+      assertThat(topic.storyList[0].chapterList[0].chapterPlayState).isEqualTo(ChapterPlayState.COMPLETED)
+      assertThat(topic.storyList[0].chapterList[1].chapterPlayState).isEqualTo(ChapterPlayState.NOT_STARTED)
+    }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testOngoingTopicList_validData_withoutAnyProgress_ongoingTopicListIsEmpty() =
```

Comment:
Is this meant to verify the topic list is empty or that it failed? The test doesn't match the test name. I'd actually expect that retrieving the ongoing list as empty when there's no progress makes sense--why would that result in a failure?

Reply:
I have written the code currently such that if the list is empty the async result fails. Should we not do this? 

Also, yes, the current name and code are not matching, I will update this once we finalise on the AsyncResult.failed is correct or not for empty list.

Reply:
Updated `createOngoingTopicListFromProgress` and `createCompletedStoryListFromProgress` implementation to return success in empty list case.